module/zfs/sa.c

```diff
@@ -451,6 +451,54 @@ sa_add_layout_entry(objset_t *os, sa_attr_type_t *attrs, int attr_count,
 	return (tb);
 }
 
+static sa_lot_t *
+sa_lookup_cached_layout(sa_os_t *sa, uint64_t hash, sa_attr_type_t *attrs,
+    int count)
+{
+	sa_layout_cache_t *slc;
+	sa_lot_t *tb = NULL;
+	int i, j;
+
+	kpreempt_disable();
+	slc = sa->sa_layout_cache + CPU_SEQID;
+
+	for (i = 0; i < SA_LAYOUT_ENTRIES; i++) {
+		if (slc->slc_lot[i] == NULL)
+			break;
+		if (slc->slc_lot[i]->lot_hash != hash)
+			continue;
+		if (sa_layout_equal(slc->slc_lot[i], attrs, count) == 0) {
+			tb = slc->slc_lot[i];
+			break;
+		}
+	}
+	if (tb != NULL) {
+		/* shift toward the most recently used */
+		j = (i + 1) % SA_LAYOUT_ENTRIES;
```

Comment:
IMHO, the most important thing is to maintain the recently used items actual, lookup itself should be cheap and simple given it's local and hit a single cacheline. again, I'm fine to fix lookup as suggested.